hack/validate/vendor

```diff
@@ -1,53 +1,53 @@
 #!/usr/bin/env bash
 
+set -e
+
 SCRIPTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 source "${SCRIPTDIR}/.validate"
 
+tidy_files=('vendor.mod' 'vendor.sum')
+vendor_files=("${tidy_files[@]}" 'vendor/')
+
+validate_vendor_tidy() {
+	# run mod tidy
+	./hack/vendor.sh tidy
+	# check if any files have changed
+	git diff --quiet HEAD -- "${tidy_files[@]}"
+}
+
 validate_vendor_diff() {
-	IFS=$'\n'
-	check_files=('vendor.sum' 'vendor.mod' 'vendor/')
-	# shellcheck disable=SC2207
-	changed_files=($(validate_diff --diff-filter=ACMR --name-only -- "${check_files[@]}" || true))
-	unset IFS
+	mapfile -t changed_files < <(validate_diff --diff-filter=ACMR --name-only -- "${vendor_files[@]}")
 
 	if [ -n "${TEST_FORCE_VALIDATE:-}" ] || [ "${#changed_files[@]}" -gt 0 ]; then
 		# recreate vendor/
-		./hack/vendor.sh
+		./hack/vendor.sh vendor
 		# check if any files have changed
-		diffs="$(git status --porcelain -- "${check_files[@]}" 2> /dev/null)"
-		mfiles="$(echo "$diffs" | awk '/^ M / {print $2}')"
-		if [ "$diffs" ]; then
-			{
-				echo 'The result of go mod vendor differs'
-				echo
-				echo "$diffs"
-				echo
-				echo 'Please vendor your package with hack/vendor.sh.'
-				echo
-				if [ -n "$mfiles" ]; then
-					git diff -- "$mfiles"
-				fi
-			} >&2
-			false
-		else
-			echo 'Congratulations! All vendoring changes are done the right way.'
-		fi
+		git diff --quiet HEAD -- "${vendor_files[@]}"
 	else
-		echo 'No vendor changes in diff.'
+		echo >&2 'No vendor changes in diff; skipping vendor check.'
 	fi
 }
 
-# 1. make sure all the vendored packages are used
-# 2. make sure all the packages contain license information (just warning, because it can cause false-positive)
-validate_vendor_used() {
-	for f in $(mawk '$1 = "#" { print $2 }' 'vendor/modules.txt'); do
-		if [ -d "vendor/$f" ]; then
-			if ! echo "vendor/$f"/* | grep -qiEc '/(LICENSE|COPYING)'; then
-				echo "WARNING: could not find copyright information for $f"
-			fi
+validate_vendor_license() {
+	while IFS= read -r module; do
+		test -d "vendor/$module" || continue
+		if ! compgen -G "vendor/$module/*" | grep -qEi '/(LICENSE|COPYING)[^/]*$'; then
+			echo >&2 "WARNING: could not find copyright information for $module"
 		fi
-	done
+	done < <(awk '/^# /{ print $2 }' vendor/modules.txt)
 }
 
-validate_vendor_diff
-validate_vendor_used
+if validate_vendor_tidy && validate_vendor_diff && validate_vendor_license; then
+	echo >&2 'Vendoring has been performed correctly!'
```

Comment:
Can you keep the (rather quirky) `Contratulations! ` prefix? While we should probably update those to be something less quirky (e.g. `PASS:` to match Go's test output), it's good to have all the `validate` scripts print the same message;


```bash
hack/make.ps1:        Write-Host "Congratulations!  All commits are properly signed with the DCO!"
hack/make.ps1:        Write-Host 'Congratulations!  ".\pkg\*.go" is safely isolated from internal code.'
hack/make.ps1:        Write-Host 'Congratulations!  All Go source files are properly formatted.'
hack/validate/changelog-date-descending:echo "Congratulations!  Changelog $changelogFile dates are in descending order."
hack/validate/changelog-well-formed:    echo "Congratulations!  Changelog $changelogFile is well-formed."
hack/validate/dco:              echo "Congratulations!  All commits are properly signed with the DCO!"
hack/validate/default-seccomp:          echo 'Congratulations!  Seccomp profile generation is done correctly.'
hack/validate/deprecate-integration-cli:        echo 'Congratulations!  No new tests were added to integration-cli.'
hack/validate/pkg-imports:      echo 'Congratulations!  Packages in "./pkg/..." are safely isolated from internal code.'
hack/validate/shfmt:    echo 'Congratulations! The shell scripts are properly formatted.'
hack/validate/swagger:          echo "Congratulations!  ${out}"
hack/validate/swagger-gen:              echo 'Congratulations!  All API changes are done the right way.'
hack/validate/toml:     echo 'Congratulations!  All TOML source files changed here have valid syntax.'
hack/validate/yamllint: echo "Congratulations! yamllint config file formatted correctly"
hack/validate/yamllint: echo "Congratulations! YAML files are formatted correctly"
```

Reply:
How do you feel about a `VALIDATE(PASS):` prefix?

Reply:
Ideally something without the `()`. Reason for that is when I search CI logs, I search for `PASS:` (or more commonly `FAIL:`)  (searching for only "fail" often gives too many results as some tests contain "fail" or some logs contain "(doesn't) fail" 😂).